src/core/unit.c

```diff
@@ -2957,13 +2957,10 @@ int unit_add_dependency(
                 return 0;
         }
 
-        if (d == UNIT_AFTER && UNIT_VTABLE(u)->refuse_after) {
-                log_unit_warning(u, "Requested dependency After=%s ignored (%s units cannot be delayed).", other->id, unit_type_to_string(u->type));
-                return 0;
-        }
-
-        if (d == UNIT_BEFORE && UNIT_VTABLE(other)->refuse_after) {
-                log_unit_warning(u, "Requested dependency Before=%s ignored (%s units cannot be delayed).", other->id, unit_type_to_string(other->type));
+        /* Note that ordering a device unit after a unit is permitted since it
+         * allows to start its job running timeout at a specific time. */
+        if (d == UNIT_BEFORE && other->type == UNIT_DEVICE) {
```

Comment:
Is there a specific reason to disallow this, given that the same dependency can be created by using After= in a device unit?

Reply:
No because `other->type != UNIT_DEVICE` in this case.

Reply:
I think my question was not phrased correctly. This check disallows a unit (say bar.service) from declaring`Before=foo.device`, while the dropped check allows foo.device to declare`After=bar.service` (d is dependency from u on other), which is essentially the same thing when translated as a reverse dependency. I've tested your patch to confirm this.

Reply:
Sorry I overlooked your initial question.

You're right that the overall result is the same with either `foo.device: After bar.service` or `bar.service: Before foo.device`.

I guess the question is now do we want to also allow user writing `bar.service: Before=foo.device` in their unit files ?

The meaning would be "start the service before the start job for the device is".

I don't really see any reason for not allowing it since we should allow its counterpart "start the device job after starting a given service" (since systemd itself generates such deps).

So yes, it seems you're right and removing the remaining check makes sense unless someone else has an objection.

Thanks.